### ISSUE TITLE
Support pack / unpack functions

### ISSUE THREAD

**Issue by iraiter2** [gpuweb/gpuweb#988]
To implement vertex pulling, we need to bind vertex buffers as flat storage buffers, for example:
[[binding 0, set 0]] var<storage_buffer> vertexBuffer0 : [[block]] struct {
  [[offset 0]] bufferData : [[stride 4]] array<u32>;
};

Then, reading from the buffer, we can obtain chunks of 4 bytes through the u32 bufferData. Currently, we would be able to read floats using "as\<f32\>(data)", and vectors by repeated reads.

The reason we need this is because attributes can refer to the same section of the buffer with different types, or have alignment problems (the alternative would be to try to generate some struct type which runs into these problems).

To be able to support vertex input formats such as "half" we need functions to unpack them, and probably include packing for consistency. These are available in GLSL std 450 under the following names:
PackSnorm4x8
PackUnorm4x8
PackSnorm2x16
PackUnorm2x16
PackHalf2x16
PackDouble2x32

UnpackSnorm4x8
UnpackUnorm4x8
UnpackSnorm2x16
UnpackUnorm2x16
UnpackHalf2x16
UnpackDouble2x32

Is there anything we need to add or would this be covered by importing GLSL.std.450? Should these be WGSL functions usable without an import?

**Comment by kvark** [comment]
Are you investigating the use of vertex pulling in the backends, or by the users of WGSL? It's not clear here.

In general, it would be great to have the packing functions in WGSL standard library.

> The reason we need this is because attributes can refer to the same section of the buffer with different types, or have alignment problems (the alternative would be to try to generate some struct type which runs into these problems).

What alignment problems? IIRC, the attribute offset has to be a multiple of its size already in WebGPU (but TODO for specification).

I think the question about collisions in the attributes has also been raised lately, and we should probably just disallow it for now. Can always switch it back to allow later.

> Is there anything we need to add or would this be covered by importing GLSL.std.450? Should these be WGSL functions usable without an import?

If you are trying to implement it in the backends, then it doesn't matter whether WGSL exposes these functions or not.

**Comment by iraiter2** [comment]
A bit of a trick, we are trying to implement vertex pulling as a WGSL->WGSL transformation. So, it would be nice to have access to these functions probably for both user and devs.

One example of problems we ran into is what should happen if an attribute has a format such as half? In GLSL without extensions shaders can use the value only after it gets converted to an f32. There wouldn't be a type for half in the shader itself anywhere since the attribute is converted, so what do we put in the structure type? It would be nice if we could bind the buffer as flat data and use these functions to do the conversion.

The main question is what relevance this has to spec stuff. We have import GLSL.std.450 in the spec text but I'm not sure if this already implies access to these functions or not, or if they should exist without importing.

**Comment by dj2** [comment]
Vertex pulling is an AST -> AST transform. We'll, usually, never write it out as WGSL.

The pack/unpack functions would get pulled in through GLSL.std.450 although note that Tint doesn't implement them.

**Comment by Kangz** [comment]
> What alignment problems? IIRC, the attribute offset has to be a multiple of its size already in WebGPU (but TODO for specification).

We have found it needs to be % 4 for Metal (although that constraint goes away with programmable pulling?)

**Comment by dneto0** [comment]
I think this needs a proposal to review. I think it should be straightforward, as it's a value-to-value transformation.
The main debating point would be namespacing, which we're already iterating on elsewere.

**Comment by dneto0** [comment]
I've started on this.
